content/blog/2020-08-19-train-a-sequence-model-with-poutyne/index.en.md

```diff
@@ -0,0 +1,563 @@
+        for word in address.split():
+            embeddings.append(self.embedding_model[word])
+        return embeddings
+     
+embedding_vectorizer = EmbeddingVectorizer()
+```
+
+We also need to apply a similar operation to the address tags (e.g. StreeNumber, StreetName). 
+This time, however, the vectorizer needs to convert the tags into categorical values (e.g. 0, 1). 
+
+For simplicity, we will use a `DatasetVectorizer` class that will apply the vectorizing process using both 
+the embedding and the address vectorize process.
+
+```python
+class DatasetVectorizer:
+    def __init__(self, embedding_vectorizer):
+        self.embedding_vectorizer = embedding_vectorizer
+        self.tags_set = {
+            "StreetNumber": 0,
+            "StreetName": 1,
+            "Unit": 2,
+            "Municipality": 3,
+            "Province": 4,
+            "PostalCode": 5,
+            "Orientation": 6,
+            "GeneralDelivery": 7
+        }
+
+    def vectorize(self, data):  # We vectorize inplace
+        for idx, item in enumerate(data):
+            data[idx] = self._item_vectorizing(item)
+
+    def _item_vectorizing(self, item):
+        address = item[0]
+        address_vector = self.embedding_vectorizer(address)
+
+        tags = item[1]
+        idx_tags = self._convert_tags_to_idx(tags)
+
+        return address_vector, idx_tags
+
+    def _convert_tags_to_idx(self, tags):
+        idx_tags = []
+        for tag in tags:
+            idx_tags.append(self.tags_set[tag])
+        return idx_tags
+
+dataset_vectorizer = DatasetVectorizer(embedding_vectorizer)
+```  
+
+Now, let's vectorize (in place) our dataset.
+
+```python
+dataset_vectorizer.vectorize(train_data)
+dataset_vectorizer.vectorize(valid_data)
+dataset_vectorizer.vectorize(test_data)
+```
+
+### DataLoader
+> We use a first trick, ``padding``.
+
+Now, since all the addresses are not of the same size, it is impossible to batch them together since all tensor elements must have the same lengths. But there is a trick, padding!
+
+The idea is simple. We add *empty* tokens at the end of each sequence up to the longest one in a batch. For the word vectors, we add vectors of 0 as padding. For the tag indices, we pad with -100s. We do so because of the [cross-entropy loss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss) and the accuracy metric all ignore targets with values of -100.
+
+To do this padding, we use the `collate_fn` argument of the [PyTorch `DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader), and on running time, that process will be done. One thing to consider, since we pad the sequence, we need each sequence's lengths to unpad them in the forward pass. That way, we can pad and pack the sequence to minimize the training time (read [this good explanation](https://stackoverflow.com/questions/51030782/why-do-we-pack-the-sequences-in-pytorch) of why we pad and pack sequences).
```

Comment:
Watch out ici je me trust pas.
```suggestion
To do this padding, we use the `collate_fn` argument of the [PyTorch `DataLoader`](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader), and on running time, that process will be done. One thing to keep in mind when treating padded sequences is that their (original) length will be required to unpad them later on, in the forward pass. That way, we can pad and pack the sequence to minimize the training time (read [this good explanation](https://stackoverflow.com/questions/51030782/why-do-we-pack-the-sequences-in-pytorch) of why we pad and pack sequences).
```

Reply:
Tu dis : That way, we can pad and pack [...]
Est-ce que le truc que tu décris juste avant (pad - unpad), c'est ça le pad and pack?

Reply:
typo c'est unpack.